src/Compilers/CSharp/Test/Semantic/Semantics/ExtensionTypeTests.cs

```diff
@@ -0,0 +1,5242 @@
+        var comp2 = CreateCompilation(lib2_cs, new[] { comp1.EmitToImageReference() });
+        comp2.VerifyDiagnostics();
+
+        var src = """
+class C<T> { }
+explicit extension R1 for UseSiteError { }
+explicit extension R2 for C<UseSiteError> { }
+class C1 : UseSiteError { }
+class C2 : C<UseSiteError> { }
+""";
+        var comp = CreateCompilation(src, new[] { comp2.EmitToImageReference() });
+        comp.VerifyDiagnostics(
+            // (2, 27): error CS0012: The type 'MissingBase' is defined in an assembly that is not referenced.You must add a reference to assembly 'missing, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null'.
+            // explicit extension R1 for UseSiteError { }
+            Diagnostic(ErrorCode.ERR_NoTypeDef, "UseSiteError").WithArguments("MissingBase", "missing, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null").WithLocation(2, 27),
+            // (4,12): error CS0012: The type 'MissingBase' is defined in an assembly that is not referenced. You must add a reference to assembly 'missing, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null'.
+            // class C1 : UseSiteError { }
+            Diagnostic(ErrorCode.ERR_NoTypeDef, "UseSiteError").WithArguments("MissingBase", "missing, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null").WithLocation(4, 12)
+            );
+    }
+
+    [Fact]
+    public void TypeDepends_SelfReference()
+    {
+        var src = """
+explicit extension R for R { }
+""";
+        var comp = CreateCompilation(src);
+        comp.VerifyDiagnostics(
+            // (1,26): error CS9205: The extended type may not be dynamic, a pointer, a ref struct, or an extension.
+            // explicit extension R for R { }
+            Diagnostic(ErrorCode.ERR_BadExtensionUnderlyingType, "R").WithLocation(1, 26)
+            );
+        var r = comp.GlobalNamespace.GetTypeMember("R");
+        Assert.Equal("R", r.ToTestDisplayString());
+        Assert.True(r.IsExtension);
+        Assert.Null(r.ExtensionUnderlyingTypeNoUseSiteDiagnostics);
+    }
+
+    [Fact]
+    public void TypeDepends_SelfReference_WithArray()
+    {
+        var src = """
+explicit extension R for R[] { }
+""";
+        var comp = CreateCompilation(src);
+        comp.VerifyDiagnostics();
+        var r = comp.GlobalNamespace.GetTypeMember("R");
+        Assert.Equal("R[]", r.ExtensionUnderlyingTypeNoUseSiteDiagnostics.ToTestDisplayString());
+    }
+
+    [Fact]
+    public void TypeDepends_SelfReference_WithTuple()
+    {
+        var src = """
+explicit extension R for (R, R) { }
+""";
+        var comp = CreateCompilation(src);
+        comp.VerifyDiagnostics();
+        var r = comp.GlobalNamespace.GetTypeMember("R");
+        Assert.Equal("(R, R)", r.ExtensionUnderlyingTypeNoUseSiteDiagnostics.ToTestDisplayString());
+    }
+
+    [Fact]
+    public void TypeDepends_SelfReference_WithStruct()
+    {
+        var src = """
+struct S<T> { }
+explicit extension R for S<R> { }
```

Comment:
FWIW, I do expect the runtime to barf on this, but it's good to have it as a test anyway :).